opm/grid/LookUpData.hh

```diff
@@ -14,51 +14,242 @@
-GNU General Public License for more details.
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
 
-You should have received a copy of the GNU General Public License
-along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #include <dune/grid/common/mcmgmapper.hh>
-#include <opm/grid/cpgrid/Entity.hpp>
+
+#include <type_traits>
+#include <vector>
 
 namespace Dune
 {
-template <typename GridType>
+class CpGrid;
+
+namespace cpgrid
+{
+template<int codim> class Entity;
+}
+
+// LookUpData class - To search data via element index
+template <typename Grid, typename GridView>
 class LookUpData
 {
 public:
-    // Constructor taking a CpGrid object
-    LookUpData(const GridType& grid) :
-        leaf_view_(grid.leafGridView()),
-        leafMapper_(leaf_view_, Dune::mcmgElementLayout())
+    // Constructor taking a GridView
+    explicit LookUpData(const  GridView& gridView) :
+        gridView_(gridView),
+        elemMapper_(gridView, Dune::mcmgElementLayout())
     {
     }
 
-    template<typename feature_type>
-    int operator()(const Dune::cpgrid::Entity<0>& elem, const std::vector<feature_type>& feature_vec)
+    // operator()(Entity, Vector) Call operator taking an EntityObject and a FeatureVector.
+    //                            Return feature of the entity, via (ACTIVE) INDEX
+    //                            For general grids, the feature vector is given for the gridView_.
+    //                            [For CpGrid, the feature vector is given for level 0.]
+    template<typename EntityType, typename FeatureType>
+    FeatureType operator()(const EntityType& elem, const std::vector<FeatureType>& feature_vec) const;
+
+    // operator()(Index, Vector)  Call operator taking an Index and a FeatureVector.
+    //                            Return feature of the entity, via (ACTIVE) INDEX
+    //                            For general grids, the feature vector is given for the gridView_.
+    //                            [For CpGrid, the feature vector is given for level 0.]
+    template<typename FeatureType>
+    FeatureType operator()(const int& elemIdx, const std::vector<FeatureType>& feature_vec) const;
+
+    // getOriginIndexFromEntity(Entity) For general grids: retunrs the same Entity index.
+    template<typename EntityType, typename GridType = Grid>
+    typename std::enable_if_t<!std::is_same_v<GridType,Dune::CpGrid>,int> getOriginIndexFromEntity(const EntityType& elem) const;
+
+    // getOriginIndexFromEntity(Entity) For CpGrid: returns index of origin cell (parent/equivalent cell when no father) in level 0.
+    template<typename EntityType, typename GridType = Grid>
```

Comment:
There seems to be no need for the template parameter GridType. You seem to be misusing it for `using GridType = Grid`. The same holds below.

Maybe we can skip it and either add a `using` directive or simply, rename the template parameter of the class or use Grid instead of GridType in `is_same_v

Reply:
I don't understand yet why this auxiliary template parameter is needed, but if I remove it, the code breaks. The error messages said "cannot overload ...". If I add it only in one method, another error shows up: "no type named ‘type’ in ‘struct std::enable_if<false, int>". 
So, for the moment, I keep this version with the extra auxiliary template parameter, coinciding with 'Grid', that allows the overloading of each method that behaves differently for CpGrid and other type of Grids.

Reply:
I see. We have discussed this privately and as the average-developer joe is not not bothered by it (no need to provide the template parameter), we can leave it as is from my side.

But I think we should do two things:
1. Add a `static_assert` that Grid is the expected type, to prevent people from using a custom grid type
2. Add documentation why it is like it is. Otherwise many people will wonder about this. This can either be done with doxygen using `\tparam` or as a regular comment about the doxygen docu.
3. Add a short comment above the class outlining that we use enable_if instead of specializing 

This would be very cool to have and help the colleagues.